common/rate-limiter/src/rate_limit.rs

```diff
@@ -0,0 +1,498 @@
+
+pub type SharedBucket = Arc<Mutex<Bucket>>;
+
+const ONE_SEC: Duration = Duration::from_secs(1);
+
+/// A generic token bucket filter
+///
+/// # Terms
+/// ## Key
+/// A `key` is an identifier of the item being rate limited
+///
+/// ## Token
+/// A `token` is the smallest discrete value that we want to rate limit by.  In a situation involving
+/// network requests, this may represent a request or a byte.  `Tokens` are the counters for the
+/// rate limiting, and when there are no `tokens` left in a `bucket`, the `key` is throttled.
+///
+/// ## Bucket
+/// A `bucket` is the tracker of the number of `tokens`.  It has a `bucket size`, and any additional
+/// tokens added to it will "spill" out of the `bucket`.  The `buckets` are filled at an `interval`
+/// with a given `fill rate`.
+///
+/// ## Interval
+/// The `interval` at which we refill *all* of the `buckets` in the token bucket filter. Configured
+/// across the whole token bucket filter.
+///
+/// ## Fill Rate
+/// The rate at which we fill a `bucket` with tokens. Configured per bucket.
+///
+/// ## Bucket Size
+/// Maximum size of a bucket.  A bucket saturates at this size.  Configured per bucket.
+///
+/// # Features
+/// ## Keys
+/// The token bucket takes any key as long as it's hashable.  This should allow it to apply to
+/// many applications that need rate limiters.
+///
+/// ## Bucket sizes and Rates
+/// ### Defaults
+/// There are defaults for bucket size and fill rate, which will apply to unknown keys.
+///
+/// ### Refill Interval
+/// Buckets are refilled automatically at an interval.  To do this synchronously, it calculates the
+/// number of intervals that have passed.  This is done synchronously and in the future may be done
+/// asynchronously.
+///
+pub struct TokenBucketRateLimiter<Key: Eq + Hash + Clone + Debug> {
+    label: &'static str,
+    buckets: RwLock<HashMap<Key, SharedBucket>>,
+    new_bucket_start_percentage: u8,
+    default_bucket_size: usize,
+    default_fill_rate: usize,
+    open: bool,
+}
+
+impl<Key: Eq + Hash + Clone + Debug> TokenBucketRateLimiter<Key> {
+    pub fn new(
+        label: &'static str,
+        new_bucket_start_percentage: u8,
+        default_bucket_size: usize,
+        default_fill_rate: usize,
+    ) -> Self {
+        // Ensure that we can actually use the rate limiter
+        assert!(new_bucket_start_percentage <= 100);
+        assert!(default_bucket_size > 0);
+        assert!(default_fill_rate > 0);
+
+        Self {
+            label,
+            buckets: RwLock::new(HashMap::new()),
+            new_bucket_start_percentage,
+            default_bucket_size,
+            default_fill_rate,
+            open: false,
+        }
+    }
+
+    pub fn test(default_bucket_size: usize, default_fill_rate: usize) -> Self {
+        Self::new("test", 100, default_bucket_size, default_fill_rate)
+    }
+
+    /// Used for testing and to not have a rate limiter
+    pub fn open(label: &'static str) -> Self {
```

Comment:
uncapped?

Reply:
I used the term `open` like an open throttle